gcc/rust/typecheck/rust-hir-type-check-pattern.cc

```diff
@@ -262,16 +262,35 @@ TypeCheckPattern::visit (HIR::TuplePattern &pattern)
 	  = *static_cast<HIR::TuplePatternItemsMultiple *> (
 	    pattern.get_items ().get ());
 
+	size_t items = ref.get_patterns ().size ();
+
+	if (parent->get_kind () != TyTy::TUPLE)
+	  {
+	    rust_error_at (pattern.get_locus (), "expected %s, found tuple",
+			   parent->as_string ().c_str ());
+	    break;
+	  }
+
+	TyTy::TupleType &par = *static_cast<TyTy::TupleType *> (parent);
+	if (ref.get_patterns ().size () != par.get_fields ().size ())
+	  {
+	    size_t fields_cnt = par.get_fields ().size ();
+	    size_t patterns_cnt = ref.get_patterns ().size ();
+	    rust_error_at (pattern.get_locus (),
+			   "expected a tuple with %lu %s, found one "
+			   "with %lu %s",
+			   (unsigned long) fields_cnt,
+			   fields_cnt == 1 ? "element" : "elements",
+			   (unsigned long) patterns_cnt,
+			   patterns_cnt == 1 ? "element" : "elements");
+	    items = std::min (items, fields_cnt);
```

Comment:
should this just return error_mark_node now that an error has occured

Reply:
This is the reason why i added the `foo` function in my test case (maybe i should include a comment to clarify). If we have something of the form

```rust
fn foo() -> i32 {
  (let b, _) = (true, 2, 3); // obv error
  use b
}
```

if we return early the error says "failed to resolve root segment", whereas if we do it as i did, we resolve the min number of matching elements (here 2) so we get that b is of type bool, and we get an error of type "expected i32, got bool".

Reply:
Perfect